duration: fix negative duration math

otan opened the pull request:
In Go, negative numbers mod positive numbers returns negative numbers
(e.g. -5 % 12 is -7), which is the root of the bug.

Resolves #52135.

Release note (bug fix): Previously, subtracting months from a
timestamp/date/duration whose date value is greater than 28
could subtract a further year off. This has now be rectified.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52138)
<!-- Reviewable:end -->

otan commented:
i don't know if it's backport worthy, but i can see some serious bugs with this.

RichardJCai commented on pkg/util/duration/duration.go:
```diff
@@ -426,6 +426,11 @@ func Add(t time.Time, d Duration) time.Time {
 	// Adjustments for 1-based math.
 	expectedMonth := time.Month((int(t.Month())-1+int(d.Months))%12 + 1)
+	// If we have a negative duration, we have a negative modulus.
+	// Round back up to the positive expectedMonth.
```
Is Round the correct term here?

RichardJCai reviewed (approved):
LGTM besides one comment nit

otan committed:
duration: fix negative duration math

In Go, negative numbers mod positive numbers returns negative numbers
(e.g. -5 % 12 is -7), which is the root of the bug.

Release note (bug fix): Previously, subtracting months from a
timestamp/date/timestamptz whose date value is greater than 28
could subtract a further year off. This has now be rectified.

otan commented:
tftr!

bors r=RichardJCai

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52138)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rohany)

___
*[pkg/util/duration/duration.go, line 429 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52138#-MDWU0Al6-zDcrk6CgDv:-MDWU0Al6-zDcrk6CgDw:b-1vzgmu) ([raw file](https://github.com/cockroachdb/cockroach/blob/229da3e8f6371afb5343fa716fc2ab6764da5b3d/pkg/util/duration/duration.go#L429)):*
> ```Go
> 	// Adjustments for 1-based math.
> 	expectedMonth := time.Month((int(t.Month())-1+int(d.Months))%MonthsPerYear) + 1
> 	// If we have a negative duration, we have a negative modulus.
> ```

Heh it's because the golang spec will tell you that `%` is remainder and not modulus

`%    remainder              integers`

https://golang.org/ref/spec#Arithmetic_operators 


<!-- Sent from Reviewable.io -->

awoods187 commented:
Can we backport this?

otan commented:
the backport PRs are ready!

otan commented:
ORMS flake!

bors r=RichardJCai
